cli/sql: invite the user to enter transactions over multiple lines.

knz opened the pull request:
This patch introduces several features and fixes a bug.
- New feature: when a user starts a transaction interactively, the
  shell now prompts the user to edit the entirety of the transaction
  locally before the SQL is sent to the server. This invites the user
  to avoid long-running transactions. The user can override this
  behavior by pressing "Enter" two times (this forces the SQL
  collected so far to go to the server).
- New feature: when the user is editing a multi-line statement (either
  a statement not yet terminated by a semicolon, or additional
  statements in a multi-line transaction), a new client-side "\show"
  command prints out the SQL code collected so far.
- Bug fix: when wrong SQL that spans multiple lines is added, the
  shell history is now populated by the individual lines of the
  erroneous statement, as opposed to a single history entry containing
  newline characters.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10187)

<!-- Reviewable:end -->

knz committed:
cli/sql: invite the user to enter transactions over multiple lines.

This patch introduces several features and fixes a bug.

- New feature: when a user starts a transaction interactively, the
  shell now prompts the user to edit the entirety of the transaction
  locally before the SQL is sent to the server. This invites the user
  to avoid long-running transactions. The user can override this
  behavior by pressing "Enter" two times (this forces the SQL
  collected so far to go to the server).

- New feature: when the user is editing a multi-line statement (either
  a statement not yet terminated by a semicolon, or additional
  statements in a multi-line transaction), a new client-side "\show"
  command prints out the SQL code collected so far.

- Bug fix: when wrong SQL that spans multiple lines is added, the
  shell history is now populated by the individual lines of the
  erroneous statement, as opposed to a single history entry containing
  newline characters.

andreimatei commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/cli/sql.go, line 149 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUytwy0cZWPNqHPnJdd:-KUytwy1lD5V6_qDwvWQ:b-h16fa8) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L149)):*
> ```Go
> // a regular query.
> func handleInputLine(
> 	ins *readline.Instance, stmt *[]string, line string, syntax parser.Syntax,
> ```

mind adding a comment about what `stmt` is?

---

*[pkg/cli/sql.go, line 186 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUytlg6It3q1hrRzYfx:-KUytlg7FvA7edAtYcWD:b-2aq8da) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L186)):*
> ```Go
> 		case `\show`:
> 			if len(*stmt) == 0 {
> 				fmt.Fprintf(osStderr, "No input so far. Did you mean SHOW?\n")
> ```

is the "did you mean" realistic? `SHOW` takes more keywords after it, right?

---

*[pkg/cli/sql.go, line 310 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUyu85hFhRSILsgZm1l:-KUyu85hFhRSILsgZm1m:b-bc3iuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L310)):*
> ```Go
> // argument ends with an incomplete transaction prefix (BEGIN without
> // ROLLBACK/COMMIT).
> func checkStatements(stmts parser.StatementList) (int, bool) {
> ```

how about `endsWithOpenTxn`?

---

*[pkg/cli/sql.go, line 310 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUyuL1KioPhvMz6wDAj:-KUyuL1LL5Fx4P1J6OOn:b-h19w4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L310)):*
> ```Go
> // argument ends with an incomplete transaction prefix (BEGIN without
> // ROLLBACK/COMMIT).
> func checkStatements(stmts parser.StatementList) (int, bool) {
> ```

mind documenting the first ret val? In fact, I don't understand what's the point of returning `len(stmts)`... The caller can do that.

---

*[pkg/cli/sql.go, line 364 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUyxKLcH7ywMTPkqDnl:-KUyxKLcH7ywMTPkqDnm:b5qfrkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L364)):*
> ```Go
> 	var stmt []string
> 	syntax := parser.Traditional
> 	partialTxnStmtCount := 0
> ```

please document this var

---

*[pkg/cli/sql.go, line 443 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUyvBzjAN9hIw1HPggY:-KUyvBzjAN9hIw1HPggZ:b250c4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L443)):*
> ```Go
> 			var parser parser.Parser
> 
> 			abstractStmts, err := parser.Parse(fullStmt, syntax)
> ```

why did you call this "abstract"? they're technically "parse trees". Anyway, to me, this name (as opposed to just `stmts` suggests that they'll be processed somehow.

---

*[pkg/cli/sql.go, line 450 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUyyjibhgM08L6H8A0B:-KUyyjibhgM08L6H8A0C:b-viojs4) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L450)):*
> ```Go
> 					break
> 				}
> 				if isInteractive {
> ```

bro I've starred a bit at this and I still can't meaningfully say anything about it. I find this function really hard to read... It's way long, way nested, has all sorts of `stmts` variables...
Would you mind breaking it up and document inputs/outputs to the sub-functions?
If you tell me there's nothing that can be done, I'll try again :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/cli/sql.go, line 149 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUytwy0cZWPNqHPnJdd:-KV1xpDSE_xtbpBwRnRy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L149)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> mind adding a comment about what `stmt` is?
</details>
Done.

---

*[pkg/cli/sql.go, line 186 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUytlg6It3q1hrRzYfx:-KV1xr1Ssp4CpoGFCnuO:b38ud1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L186)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> is the "did you mean" realistic? `SHOW` takes more keywords after it, right?
</details>
if the user enters SHOW then presses enter, they are asked to enter more syntax before they get a result. If they also enter a semicolon, they get another, more precise syntax error message from the SQL parser to help them. 

---

*[pkg/cli/sql.go, line 310 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUyu85hFhRSILsgZm1l:-KV1yComm8JGWi-wX72c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L310)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> how about `endsWithOpenTxn`?
</details>
Done.

---

*[pkg/cli/sql.go, line 310 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUyuL1KioPhvMz6wDAj:-KV1yDn7BPqVc0gd55KO:bqal5m4) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L310)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> mind documenting the first ret val? In fact, I don't understand what's the point of returning `len(stmts)`... The caller can do that.
</details>
Indeed, removed.

---

*[pkg/cli/sql.go, line 364 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUyxKLcH7ywMTPkqDnl:-KV1yFZ9U0qqMgYJufHu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L364)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> please document this var
</details>
Done.

---

*[pkg/cli/sql.go, line 443 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUyvBzjAN9hIw1HPggY:-KV1yGUKmMJMcTLJtkri:b-h86dd2) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L443)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> why did you call this "abstract"? they're technically "parse trees". Anyway, to me, this name (as opposed to just `stmts` suggests that they'll be processed somehow.
</details>
Renamed to parsedStmts.

---

*[pkg/cli/sql.go, line 450 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KUyyjibhgM08L6H8A0B:-KV1yIVnXXxAJt0IR0cp:b-nnd8bm) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea162a509c471650202e420933446366ea06bb7/pkg/cli/sql.go#L450)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> bro I've starred a bit at this and I still can't meaningfully say anything about it. I find this function really hard to read... It's way long, way nested, has all sorts of `stmts` variables...
> Would you mind breaking it up and document inputs/outputs to the sub-functions?
> If you tell me there's nothing that can be done, I'll try again :)
</details>
I've refactored the code, PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> really nice
just nits

---

Review status: 0 of 5 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/cli/sql.go, line 64 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6rjULJxvGev5nfzgO:-KV6rjUMmJsk6wpFW6Yx:b1797tf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L64)):*
> ```Go
> 	ins  *readline.Instance
> 
> 	// Options
> ```

look at this beauty

but you must have seen this nit coming - separate the "options" from the state in a `CLIConfig` object 

---

*[pkg/cli/sql.go, line 98 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6uhqBPoj_8uWEiRYR:-KV6uhqCVpXKDitbFgl3:bhx4azs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L98)):*
> ```Go
> 	// semicolon was read successfully. This is currently used for
> 	// multi-line transaction editing.
> 	partialStmtsLen int
> ```

the comment on this guy is much better than no comment at all, so you get points, but it doesn't quite tell me what this number represents. Is it the number of statements in the current txn, that haven't been sent to the server? Or is it that + whatever statements were inputed on the same line with the BEGIN?

---

*[pkg/cli/sql.go, line 102 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6vZlNfIPNF407HlS-:-KV6vZlNfIPNF407HlS0:bfckxnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L102)):*
> ```Go
> 	// concatLines is the concatenation of partialLines, computed just before
> 	// doRunStatement().
> 	concatLines string
> ```

the description sure makes it sound like we don't need this member, and it can be trivially recomputed or passed around. If that's true, I think it'd be nice to get rid of it, cause there's already a number of statements/lines-related members.

---

*[pkg/cli/sql.go, line 124 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6wDplqzXlbGZpXg3X:-KV6wDplqzXlbGZpXg3Y:b-r9qsr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L124)):*
> ```Go
> 
> const (
> 	cliStart cliStateEnum = iota // 0
> ```

are the values `// 0` important? If so, maybe it's worth a comment (cause you'd think that the whole point of typed enums is for that to not be the case). If not, get rid of them.

---

*[pkg/cli/sql.go, line 129 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6zWBF-xAqwKx60d6z:-KV6zWBF-xAqwKx60d7-:b-dsf6ti) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L129)):*
> ```Go
> 	// Querying the server for the current syntax prior to starting
> 	// input.
> 	cliQuerySyntax // 2
> ```

I like your state machine here, but aren't we going overboard? Do we really need this state, for example? We only transition to it from one other state...

---

*[pkg/cli/sql.go, line 129 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6zqGccyvcb_w9pW3d:-KV6zqGccyvcb_w9pW3e:bcg01jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L129)):*
> ```Go
> 	// Querying the server for the current syntax prior to starting
> 	// input.
> 	cliQuerySyntax // 2
> ```

What do you think of taking the opportunity to cut down on these `SHOW SYNTAX` queries? They're polluting my traces :) They're quite unnecessary... Can we only do it after a `SET` statement?

---

*[pkg/cli/sql.go, line 206 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6wvi_pBh-HpNOparF:-KV6wviaEtDOvQzSe0RO:bsfodhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L206)):*
> ```Go
> 	fmt.Fprintln(osStderr)
> 	c.exitErr = errInvalidSyntax
> 	return nextState
> ```

this function returns it's input? And doesn't even otherwise use it? Something seems fishy.
And same for the functions below... What's going on?
I kinda get it for state handlers that might decide to return a different state, but the others?

---

*[pkg/cli/sql.go, line 457 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV7-LPcx2vczukG3ljH:-KV7-LPcx2vczukG3ljI:b-tqa495) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L457)):*
> ```Go
> }
> 
> func (c *cliState) prepareStartLine(nextState cliStateEnum) cliStateEnum {
> ```

`doStartLine`, to be like your other state handlers?

---

*[pkg/cli/sql.go, line 471 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV7-jJy8KLidrGbvmQv:-KV7-jJy8KLidrGbvmQw:bqtoaoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L471)):*
> ```Go
> }
> 
> func (c *cliState) prepareContinueLine(nextState cliStateEnum) cliStateEnum {
> ```

`doContinueLine` ?

---

*[pkg/cli/sql.go, line 481 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV70hzTKEBomzHSs-Bs:-KV70hzTKEBomzHSs-Bt:b-1lp944) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L481)):*
> ```Go
> }
> 
> func (c *cliState) doReadLine(nextState cliStateEnum) cliStateEnum {
> ```

please add a comment saying that what this function does is set `lastInputLine`. The rest is noise :)
In fact, can you move the `if err=nil` case to the top and return early please

---

*[pkg/cli/sql.go, line 774 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV71ox9F3Twbz17YFg9:-KV71oxAumFfG-6qbiGx:b9fwn2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L774)):*
> ```Go
> 			state = c.doReadLine(cliDecidePath)
> 
> 		case cliDecidePath:
> ```

I'd get rid of the `cliDecidePath` state... It doesn't quite fit with the state machine model (every state "decides the path"), and it doesn't help the code. Just inline it after calling `doReadline`.

---

*[pkg/cli/sql.go, line 789 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV72TPRrcgluNwBpAsD:-KV72TPSD4Pm5oNDV2jS:b-7lvnbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L789)):*
> ```Go
> 			state = c.doProcessFirstLine(cliReadLine, cliProcessCliCmd)
> 
> 		case cliProcessCliCmd:
> ```

`cliProcesCliCmd*Maybe*` ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-:-KV74TBqSRRwAdfnNg9L:b-3ilym1)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 5 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/cli/sql.go, line 64 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6rjULJxvGev5nfzgO:-KVAx7SjPNRfEMxfmKZV:bziexwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L64)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> look at this beauty
> 
> but you must have seen this nit coming - separate the "options" from the state in a `CLIConfig` object 
</details>
But that's not right I think; this "config" is actually modified during the interactive session by \set and \unset. I though config objects are conceptually constant.

---

*[pkg/cli/sql.go, line 98 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6uhqBPoj_8uWEiRYR:-KVAxWu0nUx-Ul9zvFBR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L98)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> the comment on this guy is much better than no comment at all, so you get points, but it doesn't quite tell me what this number represents. Is it the number of statements in the current txn, that haven't been sent to the server? Or is it that + whatever statements were inputed on the same line with the BEGIN?
</details>
Done.

---

*[pkg/cli/sql.go, line 102 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6vZlNfIPNF407HlS-:-KVAy1-Gd7oJOKAPDy6K:b-uyf7ca) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L102)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> the description sure makes it sound like we don't need this member, and it can be trivially recomputed or passed around. If that's true, I think it'd be nice to get rid of it, cause there's already a number of statements/lines-related members.
</details>
Clarified in comment.

---

*[pkg/cli/sql.go, line 124 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6wDplqzXlbGZpXg3X:-KVAy9W0x12twKTv4q6C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L124)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> are the values `// 0` important? If so, maybe it's worth a comment (cause you'd think that the whole point of typed enums is for that to not be the case). If not, get rid of them.
</details>
Done.

---

*[pkg/cli/sql.go, line 129 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6zWBF-xAqwKx60d6z:-KVAyEHozg-OCYxq_fr5:b-7tqv4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L129)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I like your state machine here, but aren't we going overboard? Do we really need this state, for example? We only transition to it from one other state...
</details>
That's not correct. when one of the tests in Readline loop back they skip QuerySyntax.

---

*[pkg/cli/sql.go, line 129 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6zqGccyvcb_w9pW3d:-KVAyQGcCHyKvCPw65lF:b4hjfgn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L129)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> What do you think of taking the opportunity to cut down on these `SHOW SYNTAX` queries? They're polluting my traces :) They're quite unnecessary... Can we only do it after a `SET` statement?
</details>
This is what is being done. The SHOW SYNTAX is only issued if the previous statement contained SET.

---

*[pkg/cli/sql.go, line 206 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV6wvi_pBh-HpNOparF:-KVAyZw-ZyNHTajqMPBi:b-bdn3y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L206)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> this function returns it's input? And doesn't even otherwise use it? Something seems fishy.
> And same for the functions below... What's going on?
> I kinda get it for state handlers that might decide to return a different state, but the others?
</details>
Two reasons
1) it simplifies
```
if ... {
    c.syntaxError(...)
    return errState
} 
...
if ... {
    c.syntaxError(..)
    return errState
} 
```
into
```
if ... {
    return .syntaxError(errState, ...)
} 
...
if ... {
    return c.syntaxError(errState, ..)
} 
```

2) it is customary when you write state machine logic to chain the functions with the next state as continuation, regardless of what you do. These function simply follow the pattern, which reduces surprise. 
3) As an added bonus, if/when the Go compiler implements tail call optimization properly, this becomes faster!

---

*[pkg/cli/sql.go, line 457 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV7-LPcx2vczukG3ljH:-KVAzEvZAFuGYrvay1kD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L457)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> `doStartLine`, to be like your other state handlers?
</details>
Done.

---

*[pkg/cli/sql.go, line 471 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV7-jJy8KLidrGbvmQv:-KVAzFy-zvBTvF2KYj-6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L471)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> `doContinueLine` ?
</details>
Done.

---

*[pkg/cli/sql.go, line 481 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV70hzTKEBomzHSs-Bs:-KVAzjSaOF9Op3F09u_9:b4mw0y3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L481)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> please add a comment saying that what this function does is set `lastInputLine`. The rest is noise :)
> In fact, can you move the `if err=nil` case to the top and return early please
</details>
I added a comment.
Moving the if `err == nil` case to the top is wrong, there's a fallthrough on `err == io.EOF` too.

---

*[pkg/cli/sql.go, line 774 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV71ox9F3Twbz17YFg9:-KVB-xs503mc1LH_zaCn:b1fmee5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L774)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I'd get rid of the `cliDecidePath` state... It doesn't quite fit with the state machine model (every state "decides the path"), and it doesn't help the code. Just inline it after calling `doReadline`.
</details>
I agree this code is better in  a separate function. Done.
What you suggest is not good either, because doReadLine can return different error states other than the "normal" case, and you'd have a complex condition on the return value if you were to merge the two states in one.

---

*[pkg/cli/sql.go, line 789 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187#-KV72TPRrcgluNwBpAsD:-KVB0FFGaq8W4Op_kc2L:bul5zcb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b1680e56490226be38affc9f04d70f76b4e4bb/pkg/cli/sql.go#L789)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> `cliProcesCliCmd*Maybe*` ?
</details>
changed to `cliHandleCliCmds`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10187)*
<!-- Sent from Reviewable.io -->
